app/helpers/gtl_helper.rb

```diff
@@ -0,0 +1,125 @@
+      "type"             => "dataCtrl.gtlType === 'grid' ? 'small' : 'big'"
+    )
+  end
+
+  def miq_data_table
+    content_tag(
+      'miq-data-table', '',
+      "ng-if"            => "dataCtrl.gtlType === 'list'",
+      "ng-class"         => "{'no-action': dataCtrl.initObject.showUrl === false}",
+      "settings"         => "dataCtrl.settings",
+      "per-page"         => "dataCtrl.perPage",
+      "rows"             => "dataCtrl.gtlData.rows",
+      "on-row-click"     => "dataCtrl.onItemClicked(item, event)",
+      "on-sort"          => "dataCtrl.onSort(headerId, isAscending)",
+      "load-more-items"  => "dataCtrl.onLoadNext(start, perPage)",
+      "on-item-selected" => "dataCtrl.onItemSelect(item, isSelected)",
+      "columns"          => "dataCtrl.gtlData.cols"
+    )
+  end
+
+  # This method collects all the data comming from side channels
+  # so that `render_gtl` can be a pure function.
+  #
+  def render_gtl_outer(no_flash_div)
+    parent_id = @report_data_additional_options.try(:[], :parent_id)
+
+    options = {
+      :model_name                     => model_to_report_data,
+      :no_flash_div                   => no_flash_div || false,
+      :gtl_type_string                => @gtl_type,
+      :active_tree                    => (x_active_tree unless params[:display] || @use_action),
+      :parent_id                      => parent_id,
+      :selected_records               => gtl_selected_records,
+
+      :display                        => @display,
+      :sort_col                       => @sortcol,
+      :sort_dir                       => @sortdir,
+      :explorer                       => @explorer,
+      :view                           => @view,
+      :db                             => @db,
+
+      :report_data_additional_options => @report_data_additional_options,
+    }
+
+    render_gtl(options)
+  end
+
+  # This is a pure function. All the generated markup depends only and fully on the `options`.
+  # Assert about this function calls in your controller specs.
+  #
+  def render_gtl(options)
+    capture do
+      concat(render_gtl_markup(options[:no_flash_div]))
+      concat(render_gtl_javascripts(options))
+    end
+  end
+
+  def render_gtl_markup(no_flash_div)
+    content_tag('div', :id => 'miq-gtl-view', "ng-controller" => "reportDataController as dataCtrl") do
+      capture do
+        concat(render(:partial => 'layouts/flash_msg')) unless no_flash_div
+        concat(miq_tile_view)
+        concat(miq_data_table)
+        concat(
+          content_tag(
+            'div',
+            render(:partial => "layouts/info_msg", :locals => {:message => _("No Records Found.")}),
+            :class    => 'no-record',
+            "ng-show" => "!dataCtrl.settings.isLoading && dataCtrl.gtlData.rows.length === 0"
+          )
+        )
+      end
+    end
+  end
+
+  def render_gtl_javascripts(options)
+    parent_id_escaped = (h(j_str(options[:parent_id])) unless options[:display].nil?)
```

Comment:
imho ```h``` is no longer needed and is the default action for strings. (its used here and below)